client/state/plugins/installed/test/selectors-ts.ts

```diff
@@ -213,3 +229,108 @@ describe( 'getAllPluginsIndexedBySiteId', () => {
+	it( 'returns an object with plugins that are on the specified sites', () => {
+		const plugins = [ jetpackWithSites, akismetWithSites ];
+
+		const result = getPluginsOnSites( state, plugins );
+
+		expect( result ).toEqual( {
+			jetpack: jetpackWithSites,
+			akismet: akismetWithSites,
+		} );
+	} );
+
+	it( 'returns an empty object if the plugins have no sites', () => {
+		const plugins = [
+			{ ...jetpackWithSites, ...{ sites: {} } },
+			{ ...akismetWithSites, ...{ sites: {} } },
+		];
+
+		const result = getPluginsOnSites( state, plugins );
+
+		expect( result ).toEqual( {} );
+	} );
+} );
+
+describe( 'getPluginOnSites', () => {
+	test( 'Should get an undefined value if the requested sites are not in the current state', () => {
+		const siteIds = [ nonExistingSiteId1, nonExistingSiteId2 ];
+		expect( getPluginOnSites( state, siteIds, 'akismet' ) ).toBeUndefined();
+	} );
+
+	test( 'Should get an undefined value if the requested plugin on these sites is not in the current state', () => {
+		expect( getPluginOnSites( state, [ siteOneId ], 'jetpack' ) ).toBeUndefined();
+	} );
+
+	test( 'Should get the plugin if the it exists on one or more of the requested sites', () => {
+		const siteIds = [ siteOneId, siteTwoId ];
+		const plugin = getPluginOnSites( state, siteIds, 'hello-dolly' );
+		const sitesWithPlugins = {
+			[ siteOneId ]: {
+				active: helloDolly.active,
+				autoupdate: helloDolly.autoupdate,
+				update: helloDolly.update,
+				version: helloDolly.version,
+			},
+			[ siteTwoId ]: {
+				active: helloDolly.active,
+				autoupdate: helloDolly.autoupdate,
+				update: helloDolly.update,
+				version: helloDolly.version,
+			},
+		};
+		const { active, autoupdate, update, version, ...pluginProperties } = helloDolly;
+		expect( plugin ).toEqual( { ...pluginProperties, sites: sitesWithPlugins } );
+	} );
+} );
+
+describe( 'getPluginOnSite', () => {
+	test( 'Should get an undefined value if the requested site is not in the current state', () => {
+		expect( getPluginOnSite( state, nonExistingSiteId1, 'akismet' ) ).toBeUndefined();
+	} );
+
+	test( 'Should get an undefined value if the requested plugin on this site is not in the current state', () => {
+		expect( getPluginOnSite( state, siteOneId, 'jetpack' ) ).toBeUndefined();
+	} );
+
+	test( 'Should get the plugin if the it exists on the requested site', () => {
+		const plugin = getPluginOnSite( state, siteOneId, 'akismet' );
+		const siteProperties = {
+			active: akismet.active,
+			autoupdate: akismet.autoupdate,
+			update: akismet.update,
+			version: akismet.version,
+		};
+		const expectedPlugin = {
+			...akismet,
+			...{ sites: { [ siteOneId ]: siteProperties } },
+		};
+		expect( plugin ).toEqual( expectedPlugin );
+	} );
+} );
+
+describe( 'getPuginsOnSite', () => {
+	test( 'Should get an array of undefined if the requested site is not in the current state', () => {
+		expect( getPluginsOnSite( state, siteThreeId, [ 'jetpack' ] ) ).toEqual( [ undefined ] );
```

Comment:
That test is an example of what I asked above - doesn't it feel unexpected to have an array of `undefined`, and how are we going to be handling this at the component level? Should we just filter out the undefined values?

Reply:
In trying to minimize the scope of both this PR and the project overall I think it is best to keep the existing contracts where possible even if they are not optimal.

Reply:
The thing is that this is a behavior I'm not convinced works well in the first place. If you don't have a chance to work on it, perhaps it could be a good thing to raise it as a separate issue. The thing is that you have more context than most folks who work in the codebase. I'm fine with a follow-up, but I think we should make sure that this doesn't cause any previously uncaught unexpected behavior anywhere in the component tree.

Reply:
I looked a bit through component code uses of this selector and did find a place that depends on the `undefined` values:

https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/checkout-thank-you/marketplace/marketplace-thank-you.tsx#L93

If `undefined` is filtered out then `pluginsOnSite.every( ( pluginOnSite ) => !! pluginOnSite )` would return true every time, so `areAllPluginsOnSite` would always be `true`, and I don't think that is intended to be the case there.